### PR TITLE
Version 0.2 Tutorials

### PR DESCRIPTION
**Pull request by sslattery**
Adds tutorials for new features in version 0.2:

- Using `Kokkos::parallel_for` and `Cabana::simd_parallel_for` with `Cabana::Slice`
- Data migration with `Cabana::Distributor`
- Halo exchange with `Cabana::Halo`

The tutorials are now also added to the test suite to ensure that they both build and execute.

### CODE REVIEW

File: core/example/tutorial/13_halo_exchange/halo_exchange_example.cpp [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+        export_ranks(i) = next_rank;
+        export_ids(i) = i + num_tuple - 10;
+    }
+
+    /*
+      We have two ways to make a halo. In the first case we know where
+      we are sending the data to but not who we are receiving data from. In
+      the second we know the topology of the communication plan (i.e. who we
+      send and receive from). We know that we will only send/receive from this
+      rank and the next rank so use that information in this case.
```
Comment (dalg24) [inline core/example/tutorial/13_halo_exchange/halo_exchange_example.cpp]: State what is the motivation for doing so (doing less work when constructing `halo`)

File: core/example/tutorial/CMakeLists.txt [hunk 2]
```diff
@@ -8,6 +8,8 @@ if(Cabana_ENABLE_Serial OR Cabana_ENABLE_OpenMP OR Cabana_ENABLE_Pthread)
   add_subdirectory(07_sorting)
   add_subdirectory(08_linked_cell_list)
   add_subdirectory(09_verlet_list)
+  add_subdirectory(12_migration)
+  add_subdirectory(13_halo_exchange)
```
[the remark this answers is not in the document]
Comment (sslattery) [reply core/example/tutorial/CMakeLists.txt]: fixed

File: core/example/tutorial/11_parallel_for/parallel_for_example.cpp [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+    /*
+      Finally, perform the parallel loop. We have added a parallel for concept
+      in Cabana for this that is complementary to existing parallel for
+      implementations in Kokkos.
+
+      Note: We fence after the kernel is completed for safety but this may not
+      be needed depending on the memory/execution space being used.
+    */
+    Cabana::simd_parallel_for( simd_policy, vector_fma_kernel, "vector_fma" );
+    Kokkos::fence();
```
[the remark this answers is not in the document]
Comment (sslattery) [reply core/example/tutorial/11_parallel_for/parallel_for_example.cpp]: added a comment indicating fence is really only necessary for UVM

File: core/example/tutorial/11_parallel_for/parallel_for_example.cpp [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+      We write a Kokkos lambda function for the kernel. Note that the kernel
+      has 2 indices, s and a, corresponding to the 2D struct and array index
+      of the element on which the computation will be performed. This is
+      intended to be used with the `access()` function of the slice.
+     */
+    double v1 = 3.55;
+    double v2 = 1.24;
+    auto vector_fma_kernel =
+        KOKKOS_LAMBDA( const int s, const int a )
+        { slice_0.access(s,a) = v1 * slice_1.access(s,a) + v2; };
```
[the remark this answers is not in the document]
Comment (sslattery) [reply core/example/tutorial/11_parallel_for/parallel_for_example.cpp]: removed `v1` and `v2`

File: core/example/tutorial/12_migration/migration_example.cpp [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    int num_tuple = 100;
+    Cabana::AoSoA<DataTypes,MemorySpace,VectorLength> aosoa( num_tuple );
+
+    /*
+      Create slices and assign data. The data values are equal to id of this
+      rank so we can track where the data goes.
+     */
+    auto slice_0 = aosoa.slice<0>();
+    auto slice_1 = aosoa.slice<1>();
+    for ( int i = 0; i < num_tuple; ++i )
```
[the remark this answers is not in the document]
Comment (sslattery) [reply core/example/tutorial/12_migration/migration_example.cpp]: added a comment

File: core/example/tutorial/12_migration/migration_example.cpp [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+        export_ranks(i) = -1;
+
+    // The last 80 elements stay on this process.
+    for ( int i = 20; i < num_tuple; ++i )
+        export_ranks(i) = comm_rank;
+
+    /*
+      We have two ways to make a distributor. In the first case we know where
+      we are sending the data to but not who we are receiving data from. In
+      the second we know the topology of the communication plan (i.e. who we
```
[the remark this answers is not in the document]
Comment (sslattery) [reply core/example/tutorial/12_migration/migration_example.cpp]: fixed

File: core/example/tutorial/13_halo_exchange/halo_exchange_example.cpp (hunk 1 above)
Comment (sslattery) [reply core/example/tutorial/13_halo_exchange/halo_exchange_example.cpp]: added a comment discussing the extra global communication needed when the neighbor list is not supplied

File: core/example/tutorial/12_migration/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,3 @@
+  add_executable(Migration migration_example.cpp)
+  target_link_libraries(Migration cabanacore)
+  add_test(NAME Core_tutorial_12 COMMAND Migration)
```
Comment (dalg24) [inline core/example/tutorial/12_migration/CMakeLists.txt]:
```suggestion
  add_test(NAME Core_tutorial_12 COMMAND${MPIEXEC} ${MPIEXEC_NUMPROC_FLAG} 4 Migration)
  set_tests_properties(Core_tutorial_12 PROPERTIES PROCESSORS 4)
```

**Pull request merged**